components/scream/src/physics/p3/tests/p3_unit_tests.cpp

```diff
@@ -201,9 +201,353 @@ static void  cloud_water_autoconversion_unit_bfb_tests(){
 
 }; //  TestP3CloudWaterAutoconversion
 
-}
-}
-}
+  template <typename D>
+  struct UnitWrap::UnitTest<D>::TestP3UpdatePrognosticIce
+  {
+  
+     
+    KOKKOS_FUNCTION static void  update_prognostic_ice_unit_bfb_tests(){
+      using KTH = KokkosTypes<HostDevice>;
+      
+      static constexpr Int max_pack_size = 16;
+      
+      REQUIRE(Spack::n <= max_pack_size);
```

Comment:
You cannot use these macros inside a kokkos kernel. You need to do all the checks on host. That's because these macros can throw exceptions, and you can't do that in cuda.

Reply:
If you look at all the other codes, we do the calculations, then deep_copy the results to host, and then check the results there.

Also, for things like the check above, you could use `assert(blah)`. Asserts are fine on cuda, and this looks like a pre-condition that should be met. The `REQUIRE` macro is to check tests results.

Reply:
Wait, this is not a kernel though, right? If so, you should remove thee `KOKKOS_FUNCTION` macro. That triggers cuda compilation on GPU, which is not compatible with exceptions.

Reply:
@bartgol , i think this is not inside a kernel. @singhbalwinder did mistakenly put KOKKOS_FUNCTION on this function though.

Reply:
Easy fix then. This is what caused all the white builds to fail. At least part of it.

Reply:
Thanks @bartgol for look at the code. I tried to follow other codes in that file and all functions have `KOKKOS_FUNCTION` . Should I remove that for all of them? Let me know if I misunderstood anything.

Reply:
KOKKOS_FUNCTION is a macro that expands to `__host__ __device__` when the exec space is Cuda, and expands to nothing otherwhise. `KOKKOS_INLINE_FUNCTION` is similar, but in addition adds the `inline` keyword.

When a funciton is marked as `__device__`, it must be executable in a Cuda kernel. Part of the requirements is that such function cannot call functions that are not marked as `__device__` or throwing exceptions.

So, in short, if a function has to be executed inside a kernel, mark it as `KOKKOS_FUNCTION` or `KOKKOS_INLINE_FUNCTION`. If it's not supposed to be executed on device, do not mark it with any of those macros.

Reply:
Thanks @bartgol . I have removed that macro now.